File: world/wrappers/time_limit.py
"""world/wrappers/time_limit.py - Hard time-limit wrapper."""
from __future__ import annotations
from typing import Any, Dict, Optional, Tuple
from ..core import Env, StepResult


class TimeLimitWrapper(Env):
    """Truncates episodes after *max_steps* steps."""

    def __init__(self, env: Env, max_steps: int):
        self._env = env
        self._max_steps = max_steps
        self._elapsed = 0

    @property
    def observation_space(self):
        return self._env.observation_space

    @property
    def action_space(self):
        return self._env.action_space

    def reset(self, *, seed=None, options=None) -> Tuple[Any, Dict]:
        self._elapsed = 0
        return self._env.reset(seed=seed, options=options)

    def step(self, action) -> StepResult:
        result = self._env.step(action)
        self._elapsed += 1
        if self._elapsed >= self._max_steps and not result.terminated:
            return StepResult(result.observation, result.reward, result.terminated, True, result.info)
        return result

    def render(self, **kw):
        return self._env.render(**kw)

    def close(self):
        self._env.close()

    def __repr__(self):
        return f"TimeLimitWrapper({self._env}, max_steps={self._max_steps})"

File: world/wrappers/time_limit.ts
import { Env, StepResult } from '../core';

export class TimeLimitWrapper<ObsType, ActType> extends Env<ObsType, ActType> {
  private _env: Env<ObsType, ActType>;
  private _max_steps: number;
  private _elapsed = 0;

  get observationSpace() { return this._env.observationSpace; }
  get actionSpace() { return this._env.actionSpace; }

  constructor(env: Env<ObsType, ActType>, max_steps: number) {
    super();
    this._env = env;
    this._max_steps = max_steps;
  }

  reset(options?: { seed?: number }) {
    this._elapsed = 0;
    return this._env.reset(options);
  }

  step(action: ActType): StepResult<ObsType> {
    const result = this._env.step(action);
    this._elapsed++;
    if (this._elapsed >= this._max_steps && !result.terminated) {
      return new StepResult(result.observation, result.reward, false, true, result.info);
    }
    return result;
  }

  render() { return this._env.render(); }
  close() { this._env.close(); }
}